### DISCUSSION TITLE
How to fix nodes (other than boundary nodes) using MMG re-meshing

### DISCUSSION THREAD

**Discussion by DaveDKAF** [KratosMultiphysics/Kratos#12546] [Q&A] [answered]
Could someone please explain to me the usage of these parameters related to MMG re-meshing in the .json file

"fix_contour_model_parts"          : [],
 "fix_conditions_model_parts"       : [],
 "fix_elements_model_parts"         : [],

The information avaiable online is not the most helpful :) !  Examples of the usage would be much better.

fix_contour_model_parts: This is a list with the submodelparts of the contour where you desire to fix the nodes.

fix_conditions_model_parts : The same, but applied for conditions

fix_elements_model_parts : Idem, but in this case for elements

Many Thanks

Dave

**Comment by DaveDKAF** [comment]
Is there a parameter keyword that can be included in the json that lists the fixed nodes - would appreciate some help on this ASAP please :) !!

**Reply by loumalouomega** [reply]
A priori is `fix_contour_model_parts`. The name is inconsistent, but i added that in first place and then conditions and elements and I didn't want to break previous behaviour.

**Reply by rubenzorrilla** [reply]
Maybe we can add a `fix_contour_model_part_nodes` and start deprecating the `fix_contour_model_parts` entry.

**Reply by loumalouomega** [reply]
We could...but probably will stay deprecated forever

**Reply by sunethwarna** [reply]
Another way would be (little bit intrusive), is to set the `node.Set(Kratos.BLOCKED, True)` in pyhton to the nodes which you dont want to remesh and call the mmg process. FYI.

example use is here:
https://github.com/sunethwarna/kratos_utilities/blob/main/python/examples/test_remeshing/test_remeshing.py

**Reply by DaveDKAF** [reply]
Hi All thanks for the responses do far -  importantly I am not using python I interface the the C++ directly with Fortran; however, I think I can use your idea @sunethwarna so thanks!

**Comment by DaveDKAF** [comment]
@loumalouomega  I think it would still be very useful to have a better explanation of this though please:

A priori is fix_contour_model_parts. The name is inconsistent, but i added that in first place and then conditions and elements and I didn't want to break previous behaviour.

is the usage

"fix_contour_model_parts" : [101, 202, 303],

where nodes 101, 202, 303 are the indices of the nodes that need fixing?

**Reply by rubenzorrilla** [reply]
As far as I understand, what @loumalouomega means is that "fix_contour_model_parts" expects the name of the model parts (a string) rather than the nodal Ids.

**Reply by loumalouomega** [reply]
Indeed, it works as a list of modelparts

**Comment by DaveDKAF** [comment, marked as answer]
Maybe this is how I implement it in C++?

for (auto& node : r_Model_Part.Nodes()) {

         if (node.Id() == 101) {
         node.Set(BLOCKED, true);
         }
        }

**Comment by DaveDKAF** [comment]
So, I'll answer my own question  - I can confirm that that works in C++ :)  Thanks all.

Dave
